File: Python_Ver/main_mannualDecoder.py
import os
import numpy as np
from utils import data_setup
from utils.opt_Setup_real import opt_Setup_real
from utils.PreProcess import PreProcess
from training import RLPP
from scipy.io import loadmat

# Ensure the results directory exists
if not os.path.exists('./Python_Ver/results'):
    os.makedirs('./Python_Ver/results')

# Model training
DataName = "./Python_Ver/data/rat02.mat"

data = data_setup(DataName)  # Load data
data["M1num_pre"] = 4

data["modelName"] = "decodingModel_manual"
data["his"] = 50
data["DataIndex"] = data["DataIndex"] + "Manual"
print(data["DataIndex"])

opt = opt_Setup_real(data)  # Set options

resultsDir = f'./Python_Ver/results/{opt["DataIndex"]}/'
if not os.path.exists(resultsDir):
    os.makedirs(resultsDir)

np.random.seed(0)  # Set random seed to default

# PreProcess: get mPFC ensemble for every time bin
inputEnsemble, M1Truth, Actions, Trials, opt = PreProcess(
    data["mPFC"], data["M1"], data["segment"], data["trialNo"], opt["decay_parameter"], opt
)

sch = "RL"

for testFold in range(opt["foldNum"]):  # 5-fold cross-validation
    opt["testFold"] = testFold
    trainFolds = list(range(opt["foldNum"]))
    trainFolds.remove(testFold)
    opt["trainTrials"] = np.concatenate([opt["folds"][i] for i in trainFolds]).flatten()
    opt["NumberOfTrainTrials"] = len(opt["trainTrials"])
    opt["testTrials"] = opt["folds"][testFold]
    opt["NumberOfTestTrials"] = len(opt["testTrials"])

    print(f'----{opt["DataIndex"]} {sch} fold {opt["testFold"]} Train start----')
    
    for i in range(opt["ReTrainTimes"]):
        RLPP(data, opt, inputEnsemble, M1Truth, Actions, Trials)

    # Evaluate results
    paths = [f for f in os.listdir(resultsDir) if f.startswith(f'{sch}_{opt["testFold"]}_') and f.endswith('.mat')]
    SucRate = []
    testSuc = []
    
    for path in paths:
        file_path = os.path.join(resultsDir, path)
        data_dict = loadmat(file_path)
        SucRate.append(data_dict["MaxReward"].item())
        testSuc.append(data_dict["testSucRate"].item())
    
    best_idx = np.argmax(SucRate)
    best_path = paths[best_idx]
    source = os.path.join(resultsDir, best_path)
    destination = f'./Python_Ver/results/MATdata_{data["DataIndex"]}_{sch}_{opt["testFold"]}.mat'
    
    os.replace(source, destination)
    print(f'{destination} testSuc: {testSuc[best_idx]}')

File: main_mannualDecoder.m
clc;clear;close all;
addpath model/ utils/ decoding/ training/

if ~exist('./results', 'dir')
    mkdir('./results');
end

%% model training
DataName = "data/rat02.mat";

data = data_setup(DataName); % load data
data.M1num_pre = 4;
data.modelName = 'decodingModel_manual';
dta.his = 50;
data.DataIndex = [data.DataIndex, 'Manual'];

opt = opt_Setup_real(data); % set options

if ~exist(['./results/', opt.DataIndex,'/'], 'dir')
    mkdir(['./results/', opt.DataIndex,'/']);
end

rng('default');
% PreProcess: get mPFC ensemble for every time bin
[inputEnsemble,M1_truth,Actions,Trials,opt] = PreProcess( ...
    data.mPFC,data.M1,data.segment,data.trialNo,opt.decay_parameter,opt);

sch = 'RL';

for testFold = 1:opt.foldNum  % explore testFold in 1:5 for 5-fold cross validation
    opt.testFold = testFold;
    trainFolds = 1:opt.foldNum; trainFolds(testFold) = [];
    opt.trainTrials = cell2mat(opt.folds(trainFolds));
    opt.NumberOfTrainTrials = length(opt.trainTrials);
    opt.testTrials = opt.folds{testFold};
    opt.NumberOfTestTrials = length(opt.testTrials);

    disp(['----', opt.DataIndex,' ',sch,' fold ',num2str(opt.testFold),' Train start----']);
    parfor i = 1:opt.ReTrainTimes
        s = rng;
        RLPP(data,opt,s,inputEnsemble,M1_truth,Actions,Trials);
    end

    paths = dir(['results/',data.DataIndex,'/',sch,'_',num2str(opt.testFold),'_','*.mat']);
    SucRate = zeros(1,length(paths));
    testSuc = zeros(1,length(paths));
    for i=1:length(paths)
        load(['results/',data.DataIndex,'/',paths(i).name],"MaxReward","testSucRate")
        SucRate(i) = MaxReward;
        testSuc(i) = testSucRate;
    end
    [best,I] = max(SucRate);
    source = ['results/',data.DataIndex,'/',paths(I).name];
    destination = ['results/', data.DataIndex, '_', sch, '_', num2str(opt.testFold), '.mat'];
    copyfile(source, destination, 'f')
    disp([destination, ' testSuc: ', num2str(testSuc(I))])
end
